File: Sample Code/ObjcSampleCode/DJISdkDemo/Demo/FlightController/IntelligentAssistant/FCIntelligentAssistantViewController.m
//
//  FCIntelligentAssistantViewController.m
//  DJISdkDemo
//
//  Copyright © 2016 DJI. All rights reserved.
//

#import "FCIntelligentAssistantViewController.h"
#import <DJISDK/DJISDK.h>
#import "DemoUtility.h"


@interface FCIntelligentAssistantViewController () <DJIFlightAssistantDelegate>

@property (weak, nonatomic) IBOutlet UISwitch *collisionAvoidanceEnable;
@property (weak, nonatomic) IBOutlet UISwitch *visionPositioningEnable;
@property (weak, nonatomic) IBOutlet UILabel *isSensorWorking;
@property (weak, nonatomic) IBOutlet UILabel *isBraking;
@property (weak, nonatomic) IBOutlet UILabel *systemWarning;

@property (weak, nonatomic) IBOutlet UILabel *l2Distance;
@property (weak, nonatomic) IBOutlet UILabel *l2WarningLevel;

@property (weak, nonatomic) IBOutlet UILabel *l1Distance;
@property (weak, nonatomic) IBOutlet UILabel *l1WarningLevel;

@property (weak, nonatomic) IBOutlet UILabel *r1Distance;
@property (weak, nonatomic) IBOutlet UILabel *r1WarningLevel;

@property (weak, nonatomic) IBOutlet UILabel *r2Distance;
@property (weak, nonatomic) IBOutlet UILabel *r2WarningLevel;

@end

@implementation FCIntelligentAssistantViewController

-(void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    DJIFlightController* fc = [DemoComponentHelper fetchFlightController];
    if (!fc || !fc.flightAssistant) {
        ShowResult(@"Flight controller or flight assistant is not detected. ");
        return;
    }
    
    [fc.flightAssistant setDelegate:self];
    [self updateFlightAssistantSwitches];
}

-(void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    
    DJIFlightController* fc = [DemoComponentHelper fetchFlightController];
    if (fc && fc.flightAssistant) {
        [fc.flightAssistant setDelegate:nil];
    }
}


- (IBAction)onCollisionAvoidanceSwitchValueChanged:(id)sender {
    DJIFlightController* fc = [DemoComponentHelper fetchFlightController];
    if (fc && fc.flightAssistant) {
        [fc.flightAssistant setCollisionAvoidanceEnabled:((UISwitch*)sender).on withCompletion:^(NSError * _Nullable error) {
            if (error) {
                ShowResult(@"Set collision avoidance enabled failed: %@", error.description);
            }
        }];
    }
}

- (IBAction)onVisionPositioningSwitchValueChanged:(id)sender {
    DJIFlightController* fc = [DemoComponentHelper fetchFlightController];
    if (fc) {
        [fc setVisionAssistedPositioningEnabled:((UISwitch*)sender).on withCompletion:^(NSError * _Nullable error) {
            if (error) {
                ShowResult(@"Set vision positioning enabled failed: @%", error.description);
            }
        }];
    }
}

- (void)updateFlightAssistantSwitches {
    DJIFlightController* fc = [DemoComponentHelper fetchFlightController];
    DJIFlightAssistant* assistant = fc.flightAssistant;
    WeakRef(target);
    [assistant getCollisionAvoidanceEnabledWithCompletion:^(BOOL enable, NSError * _Nullable error) {
        WeakReturn(target);
        target.collisionAvoidanceEnable.on = enable;
    }];

    [fc getVisionAssistedPositioningEnabledWithCompletion:^(BOOL enabled, NSError * _Nullable error) {
        WeakReturn(target);
        target.visionPositioningEnable.on = enabled;
    }];
}

-(void)flightAssistant:(DJIFlightAssistant *)assistant didUpdateVisionDetectionState:(DJIVisionDetectionState *)state
{
    if (state.position != DJIVisionSensorPositionNose) {
        return; 
    }
    self.isSensorWorking.text = state.isSensorBeingUsed ? @"YES" : @"NO";
    self.systemWarning.text = [self stringWithSystemWarning:state.systemWarning];
    
    self.l2Distance.text = [NSString stringWithFormat:@"%f",
                            ((DJIObstacleDetectionSector *)(state.detectionSectors[0])).obstacleDistanceInMeters];
    self.l2WarningLevel.text = [self stringWithSector:
                                ((DJIObstacleDetectionSector *)(state.detectionSectors[0])).warningLevel];
    
    self.l1Distance.text = [NSString stringWithFormat:@"%f",
                            ((DJIObstacleDetectionSector *)(state.detectionSectors[1])).obstacleDistanceInMeters];
    self.l1WarningLevel.text = [self stringWithSector:
                                ((DJIObstacleDetectionSector *)(state.detectionSectors[1])).warningLevel];
    
    self.r1Distance.text = [NSString stringWithFormat:@"%f",
                            ((DJIObstacleDetectionSector *)(state.detectionSectors[2])).obstacleDistanceInMeters];
    self.r1WarningLevel.text = [self stringWithSector:
                                ((DJIObstacleDetectionSector *)(state.detectionSectors[2])).warningLevel];
    
    self.r2Distance.text = [NSString stringWithFormat:@"%f",
                            ((DJIObstacleDetectionSector *)(state.detectionSectors[3])).obstacleDistanceInMeters];
    self.r2WarningLevel.text = [self stringWithSector:
                                ((DJIObstacleDetectionSector *)(state.detectionSectors[3])).warningLevel];
}

-(NSString*) stringWithSystemWarning:(DJIVisionSystemWarning) warning {
    switch (warning) {
        case DJIVisionSystemWarningInvalid:
            return @"Invalid";
            
        case DJIVisionSystemWarningSafe:
            return @"Safe";
        
        case DJIVisionSystemWarningDangerous:
            return @"Dangerous";
            
        case DJIVisionSystemWarningUnknown:
            return @"Unknown";
            
        default:
            break;
    }
    return @"";
}

-(NSString*) stringWithSector:(DJIObstacleDetectionSectorWarning) warning {
    switch (warning) {
        case DJIObstacleDetectionSectorWarningInvalid:
            return @"NA";
            
        case DJIObstacleDetectionSectorWarningLevel1:
            return @"1";
            
        case DJIObstacleDetectionSectorWarningLevel2:
            return @"2";
            
        case DJIObstacleDetectionSectorWarningLevel3:
            return @"3";
        case DJIObstacleDetectionSectorWarningLevel4:
            return @"4";
            
        default:
            return @"XX";
    }
}

@end

File: Sample Code/SwiftSampleCode/DJISDKSwiftDemo/Components/FlightController/IntelligentAssistant/FCIntelligentAssistantViewController.swift
//
//  FCIntelligentAssistantViewController.swift
//  DJISDKSwiftDemo
//
//  Created by DJI on 16/1/5.
//  Copyright © 2016 DJI. All rights reserved.
//

import DJISDK
class FCIntelligentAssistantViewController: DJIBaseViewController, DJIIntelligentFlightAssistantDelegate {
    
    @IBOutlet weak var isSensorWorking: UILabel!
    @IBOutlet weak var isBraking: UILabel!
    @IBOutlet weak var systemWarning: UILabel!
    @IBOutlet weak var collisionAvoidanceEnable: UISwitch!
    @IBOutlet weak var visionPositioningEnable: UISwitch!
    // Sectors:
    @IBOutlet weak var l2Distance: UILabel!
    @IBOutlet weak var l2WarningLevel: UILabel!
    @IBOutlet weak var l1Distance: UILabel!
    @IBOutlet weak var l1WarningLevel: UILabel!
    @IBOutlet weak var r1Distance: UILabel!
    @IBOutlet weak var r1WarningLevel: UILabel!
    @IBOutlet weak var r2Distance: UILabel!
    @IBOutlet weak var r2WarningLevel: UILabel!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view from its nib.
        let fc: DJIFlightController? = self.fetchFlightController()
        if (fc != nil && fc?.intelligentFlightAssistant != nil) {
            fc?.intelligentFlightAssistant?.delegate = self
            updateSwitchState(fc?.intelligentFlightAssistant);
        }
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    func updateSwitchState(_ fa:DJIIntelligentFlightAssistant?){
        if (fa != nil) {
            fa?.getCollisionAvoidanceEnabled(completion: { (state:Bool, error: Error?) -> Void in
                if (error != nil) {
                    self.showAlertResult("Error to get collision avoidance:\(error?.localizedDescription)")
                }else {
                    self.collisionAvoidanceEnable.setOn(state, animated:false)
                }
                
                fa?.getVisionPositioningEnabled(completion: { (state:Bool, error:Error?) -> Void in
                    if (error != nil) {
                        self.showAlertResult("Error to get vision positioning:\(error?.localizedDescription)")
                    }else {
                        self.visionPositioningEnable.setOn(state, animated:false)
                    }
                })
            })
        }
    }
    
    func intelligentFlightAssistant(_ assistant: DJIIntelligentFlightAssistant, didUpdate state: DJIVisionDetectionState) {
        
        self.isSensorWorking.text = state.isSensorWorking.description
        self.systemWarning.text = stringWithSystemWarnings(state.systemWarning)
        
        let firstSector : DJIVisionDetectionSector? = state.detectionSectors?[0]
        self.l2Distance.text = firstSector?.obstacleDistanceInMeters.description
        self.l2WarningLevel.text = firstSector?.warningLevel.rawValue.description
        
        self.l1Distance.text = (state.detectionSectors?[1])?.obstacleDistanceInMeters.description
        self.l1WarningLevel.text = (state.detectionSectors?[1])?.warningLevel.rawValue.description

        self.r1Distance.text = (state.detectionSectors?[2])?.obstacleDistanceInMeters.description
        self.r1WarningLevel.text = (state.detectionSectors?[2])?.warningLevel.rawValue.description
        
        self.r2Distance.text = (state.detectionSectors?[3])?.obstacleDistanceInMeters.description
        self.r2WarningLevel.text = (state.detectionSectors?[3])?.warningLevel.rawValue.description
    }
    
    func stringWithSystemWarnings(_ status: DJIVisionSystemWarning) -> String {
        if status == DJIVisionSystemWarning.unknown {
            return "Unknown"
        }
        else if status == DJIVisionSystemWarning.invalid {
            return "Invalid"
        }
        else if status == DJIVisionSystemWarning.safe {
            return "Safe"
        }
        else if status == DJIVisionSystemWarning.dangerous {
            return "Dangerous"
        }
        
        return "SDK Wrong"
        
    }
    
    @IBAction func onCollisionAvoidanceSwitchValueChanged(_ sender: UISwitch){
        let fc: DJIFlightController? = self.fetchFlightController()
        let fa = fc?.intelligentFlightAssistant
        if (fa != nil) {
          fa?.setCollisionAvoidanceEnabled(sender.isOn, withCompletion: { (error:Error?) -> Void in
            if (error != nil) {
                self.showAlertResult("Error to enable/disable CollisionAvoidance:\(error)")
                sender.setOn(!sender.isOn, animated:false)
            }
          })
        }
    }
    
    @IBAction func onVisionPositioningSwitchValueChanged(_ sender: UISwitch){
        let fc: DJIFlightController? = self.fetchFlightController()
        let fa = fc?.intelligentFlightAssistant
        if (fa != nil) {
            fa?.setVisionPositioningEnabled(sender.isOn, withCompletion: { (error:Error?) -> Void in
                if (error != nil) {
                    self.showAlertResult("Error to enable/disable VisionPositioning:\(error)")
                    sender.setOn(!sender.isOn, animated:false)
                }
            })
        }
    }
    
}
